File: c++/MaybeHug.cpp
#include <iostream>
#include <typeinfo>
#include "Cutie.h"
#include "MaybeHug.h"

MaybeHug::MaybeHug(Cutie* cutie)
{
	if (cutie == NULL)
	{
		throw NoCutieException();
	}
	this->cutie = cutie;
}

void MaybeHug::hug()
{
	try
	{
		if (cutie->acceptsHugs())
		{
			cutie->hug();
		}
		else
		{
			cutie->empathy();
		}
	}
	catch (NoCutieException)
	{
		std::cerr << "No Cutie provided to MaybeHug :(";
	}
}

File: java/MaybeHug.java
public class MaybeHug {

    private Cutie cutie;

    public MaybeHug(Cutie cutie) throws NoCutieException {
        if (cutie == null) {
            throw new NoCutieException();
        }
        this.cutie = cutie;
    }

    public void hug() {
        if (cutie.acceptsHugs()) {
            cutie.hug();
        } else {
            cutie.empathy();
        }
    }
}
